File: CS/RuntimeDbChooser.Web/WebApplicationEx.cs
using System;
using System.Web;
using System.Web.Security;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Web;
using RuntimeDbChooser.Module.Web;
using RuntimeDbChooser.Module.BusinessObjects;

namespace RuntimeDbChooser.Web {
	public partial class RuntimeDbChooserAspNetApplication : WebApplication {
		protected override void ReadSecuredLogonParameters() {
			base.ReadSecuredLogonParameters(); // the "UserName" is restored in the base method.

			string databaseName = HttpContext.Current.Request.Params[WebChangeDatabaseController.DatabaseParameterName];
			if(!string.IsNullOrEmpty(databaseName)) {
				((IDatabaseNameParameter)SecuritySystem.LogonParameters).DatabaseName = databaseName;
			}
		}
		private bool canReadSecuredLogonParameters = true;
		protected override bool CanReadSecuredLogonParameters() {
			if(!canReadSecuredLogonParameters)
				return false;
			return base.CanReadSecuredLogonParameters();
		}
		protected override bool OnLogonFailed(object logonParameters, Exception e) {
			if(CanReadSecuredLogonParameters()) {
				FormsAuthentication.SignOut();
				canReadSecuredLogonParameters = false;
				try {
					Start();
				} finally {
					canReadSecuredLogonParameters = true;
				}
				return true;
			}
			else {
				return base.OnLogonFailed(logonParameters, e);
			}
		}
		protected override void ReadLastLogonParametersCore(DevExpress.ExpressApp.Utils.SettingsStorage storage, object logonObject) {
			//string databaseName = HttpContext.Current.Request.Params[WebChangeDatabaseController.DatabaseParameterName];
			if(string.IsNullOrEmpty(((IDatabaseNameParameter)logonObject).DatabaseName)) {
				base.ReadLastLogonParametersCore(storage, logonObject);
			}
		}
		protected override void OnLoggingOn(LogonEventArgs args) {
			base.OnLoggingOn(args);
			MSSqlServerChangeDatabaseHelper.UpdateDatabaseName(this, ((IDatabaseNameParameter)args.LogonParameters).DatabaseName);
		}
	}
}

File: VB/RuntimeDbChooser.Web/WebApplicationEx.vb
Imports System
Imports System.Web
Imports System.Web.Security
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Web
Imports RuntimeDbChooser.Module.Web
Imports RuntimeDbChooser.Module.BusinessObjects

Namespace RuntimeDbChooser.Web
    Partial Public Class RuntimeDbChooserAspNetApplication
        Inherits WebApplication

        Protected Overrides Sub ReadSecuredLogonParameters()
            MyBase.ReadSecuredLogonParameters() ' the "UserName" is restored in the base method.

            Dim databaseName As String = HttpContext.Current.Request.Params(WebChangeDatabaseController.DatabaseParameterName)
            If Not String.IsNullOrEmpty(databaseName) Then
                DirectCast(SecuritySystem.LogonParameters, IDatabaseNameParameter).DatabaseName = databaseName
            End If
        End Sub

        Private canReadSecuredLogonParameters_Renamed As Boolean = True
        Protected Overrides Function CanReadSecuredLogonParameters() As Boolean
            If Not canReadSecuredLogonParameters_Renamed Then
                Return False
            End If
            Return MyBase.CanReadSecuredLogonParameters()
        End Function
        Protected Overrides Function OnLogonFailed(ByVal logonParameters As Object, ByVal e As Exception) As Boolean
            If CanReadSecuredLogonParameters() Then
                FormsAuthentication.SignOut()
                canReadSecuredLogonParameters_Renamed = False
                Try
                    Start()
                Finally
                    canReadSecuredLogonParameters_Renamed = True
                End Try
                Return True
            Else
                Return MyBase.OnLogonFailed(logonParameters, e)
            End If
        End Function
        Protected Overrides Sub ReadLastLogonParametersCore(ByVal storage As DevExpress.ExpressApp.Utils.SettingsStorage, ByVal logonObject As Object)
            'string databaseName = HttpContext.Current.Request.Params[WebChangeDatabaseController.DatabaseParameterName];
            If String.IsNullOrEmpty(DirectCast(logonObject, IDatabaseNameParameter).DatabaseName) Then
                MyBase.ReadLastLogonParametersCore(storage, logonObject)
            End If
        End Sub
        Protected Overrides Sub OnLoggingOn(ByVal args As LogonEventArgs)
            MyBase.OnLoggingOn(args)
            MSSqlServerChangeDatabaseHelper.UpdateDatabaseName(Me, DirectCast(args.LogonParameters, IDatabaseNameParameter).DatabaseName)
        End Sub
    End Class
End Namespace